packages/shared/src/telemetry/collector.ts

```diff
@@ -119,13 +123,23 @@ export class TelemetryCollector {
     this.#scheduleFlush();
   }
 
-  #shouldRecord(eventSamplingRate?: number): boolean {
+  #shouldRecord(event: TelemetryEventRaw) {
+    return this.isEnabled && !this.isDebug && this.#shouldBeSampled(event);
+  }
+
+  #shouldBeSampled(event: TelemetryEventRaw) {
     const randomSeed = Math.random();
-    const shouldBeSampled =
-      randomSeed <= this.#config.samplingRate &&
-      (typeof eventSamplingRate === 'undefined' || randomSeed <= eventSamplingRate);
+    const clientCache = this.#clientCache;
 
-    return this.isEnabled && !this.isDebug && shouldBeSampled;
+    if (clientCache.isStorageSupported) {
+      const isCached = clientCache.cacheAndRetrieve(event);
+      return !isCached;
+    }
```

Comment:
Can we include the `isStorageSupported` check in `cacheAndRetrieve()`? That would simplify usage here